### ISSUE TITLE
Local Function doesn't update null state

### ISSUE THREAD

**Issue by YairHalberstadt** [dotnet/roslyn#41898]
**Version Used**:  master as of 16th feb 2020

**Steps to Reproduce**:

Example 1.

```csharp
using System;
public class C {
    #nullable enable
    public void M() {
        string? x = "";
        LocalF();
        Console.WriteLine(x.ToString());
        void LocalF() => x = null;
    }
}
```

**Expected Behavior**: Warning on `x.ToString()`

**Actual Behavior**: no warning

Example 2.

```csharp
using System;
public class C {
    #nullable enable
    public void M() {
        string? x = null;
        LocalF();
        Console.WriteLine(x.ToString());
        void LocalF() => x = "";
    }
}
```

**Expected Behavior**: No warning

**Actual Behavior**: Warning on `x.ToString()`

**Comment by jcouv** [comment]
Tagging @agocke for triage

**Comment by agocke** [comment]
This is by design. The design for local functions is

1. The null state at the start of a local function is the intersection of all of the states where it is called.

2. The null state after a call to a local function is unchanged.

This is different from definite assignment, and the reasoning is pretty simple. In definite assignment, you can't observe the flow state except by using variables. For nullable reference types, however, the flow state is part of the inferred types of variables, so it is observable through the semantic model and can't be flow-dependent.

I discarded other designs as either being too complicated or simply too computationally expensive.
